Use the Logos Chat module API from an app (#365)
status-im/ift-docs: docs/messaging/chat-module/build-logos-module-that-uses-chat-module-api.md

---
title: Build a Logos module that uses the Chat module API
doc_type: procedure
product: messaging

> topics: []

Reviewer: `authors` and `topics` are both empty here — the Delivery doc fills them in (`authors: igor-sirotin, kashepavadan`, `topics: delivery`). Mind adding yourself as author and a topic like `chat`?

   ```

## Step 2: Declare `chat_module` as a dependency

Add `chat_module` to both `metadata.json` and `flake.nix`, pinning to the released tag so your app stays stable as the module's API evolves. Because `chat_module` now reaches the delivery node over IPC, also declare its runtime transport dependency, `delivery_module`, and reuse the `delivery_module` contract that `chat_module` bundles.

{% hint style="info" %}
The flake input name (`chat_module`) must match the dependency name in `metadata.json`. `logos-module-builder` automatically generates the typed `chat_module` wrapper at build time.
{% endhint %}

1. In `metadata.json`, declare the dependencies and reuse `chat_module`'s bundled `delivery_module` contract:

   ```json
   {
     "name": "my_app",
     "dependencies": ["chat_module", "delivery_module"],
     "dependency_overrides": {
       "delivery_module": {
         "input": "chat_module",
         "file": "rust-lib/deps/delivery_module.lidl"
       }
     },
     ...
   }
   ```

1. In `flake.nix`, add a matching pinned input:

   ```nix
   inputs = {
     logos-module-builder.url = "github:logos-co/logos-module-builder";

>      chat_module.url = "github:logos-co/logos-chat-module/v0.1.1";

Reviewer: **(blocker) This snippet is missing a piece and the build won't resolve `delivery_module`.**

`metadata.json` lists `delivery_module` as a dependency, and the module builder pulls each dependency's runtime from a matching flake input — but here the only input is `chat_module`. The `dependency_overrides` entry only points the builder at delivery's `.lidl` contract for codegen; it doesn't supply the runtime build. So `nix build` has nothing to resolve `delivery_module` against.

`logos-chat-ui` handles this by pinning delivery separately and passing it in:

```nix
logos-delivery-module.url = "github:logos-co/logos-delivery-module/<rev>";
# then: flakeInputs = { delivery_module = logos-delivery-module; } // inputs;
```

Could we add that to the snippet? One heads-up while you're in there: `logos-chat-ui` deliberately pins a commit *past* `v0.1.2` to pick up the zerokit/RLN build fix (the `v0.1.2` tag trips over a crates.io 403 during the Rust build). Worth a warning so people don't pin the tag and get stuck.

   });
   // One of your own messages was recorded/sent.
   chat.on("message_sent", [](const QVariantList& a) {
       // a[0]: convo_id, a[1]: content, a[2]: timestamp_ms
   });
   // A conversation was created — incoming from a peer, or your own outgoing one.
   chat.on("conversation_created", [](const QVariantList& a) {
       // a[0]: QString convo_id, a[1]: bool is_outgoing, a[2]: QString peer_label
   });
   chat.on("conversation_updated", [](const QVariantList& a) { /* a[0]: convo_id */ });
   chat.on("conversation_deleted", [](const QVariantList& a) { /* a[0]: convo_id */ });
   // Delivery/connection state changed — drives your "connected" indicator.
   chat.on("delivery_state_changed", [](const QVariantList& a) {
       // a[0]: QString delivery_state ("initialising" | "online" | "stopped" | "error")
       // a[1]: QString detail
   });
   ```

   - Events push over the IPC event channel automatically — there is no separate subscribe/enable call. See [`rust-lib/chat_module.lidl`](https://github.com/logos-co/logos-chat-module/blob/v0.1.1/rust-lib/chat_module.lidl) for the exact argument list of every method and event.

## Step 4: Drive the chat lifecycle

> Status-bearing methods return their result **synchronously** as a `LogosResult`: `res.success` tells you whether the call succeeded, `res.getError<QString>()` carries the failure reason, and `res.getValue<QString>()` carries the returned value (for example, the intro bundle). Ongoing activity — incoming messages, new conversations, delivery-state changes — arrives asynchronously through the push events you subscribed to in Step 3.

Reviewer: Small thing that'll bite copy-pasters: here we use `res.getError<QString>()` / `res.getValue<QString>()`, but the Delivery API doc uses `r.getError()` / `r.getString()` for the same `LogosResult` type. Someone reading both will hit a compile error on one of them. Do you know which form the SDK actually exposes? Would be good to line the two docs up.

{% hint style="info" %}
`init()` starts delivery asynchronously, so the client is not connected the moment `init()` returns. Watch `delivery_state_changed` (or poll `status()`) for the `online` state before creating conversations or sending messages.
{% endhint %}

`init()` takes the instance directory, a delivery preset, and a TCP port:

| Parameter | Type | Notes |
|---|---|---|
| `instance_path` | string | Directory for this instance's persistent state (`identity.db`, `history.json`). Use a distinct directory per instance to run several side by side. |
| `delivery_preset` | string | Network preset for the delivery node. Use `logos.test` to reach the Logos test network. Must match across all participants. |

> | `tcp_port` | int | Logos Delivery (Waku) TCP port. `0` picks a random port. |

Reviewer: (nit, totally optional) the `tcp_port` note calls it the "Logos Delivery (Waku) TCP port." If the docs generally keep Waku under the hood, might want to drop it; if not, ignore me.

1. Initialise the chat client:

   ```cpp
   const LogosResult res = m_logos->chat_module.init("/path/to/instance", "logos.test", 60000);
   if (!res.success) {
       qWarning() << "init failed:" << res.getError<QString>();
       return;
   }
   // init succeeded; delivery connects asynchronously.
   // Wait for delivery_state_changed with state == "online" before creating conversations.
   ```

1. Read (or set) your identity:

   ```cpp
   const QString myId = m_logos->chat_module.get_installation_name();
   // Optionally choose a name: m_logos->chat_module.set_installation_name("alice");
   ```

1. Create and share your introduction bundle:

   ```cpp
   const LogosResult bundle = m_logos->chat_module.create_intro_bundle();
   if (bundle.success)
       shareOutOfBand(bundle.getValue<QString>());   // the intro bundle string
   ```

1. Open a private conversation as the initiator, or receive one as the recipient:

   ```cpp
   // content is plain text — no encoding required.

>    const LogosResult res = m_logos->chat_module.create_conversation(peerBundle, "Hello!");

Reviewer: Minor: the contract also has `set_conversation_nickname` and `delete_conversation`, and they're what fire the `conversation_updated` / `conversation_deleted` events the doc already covers. A one-liner pointing at them would help anyone who wants to rename or delete a conversation.

   // On success a conversation_created event (is_outgoing == true) fires with the new convo_id.
   ```

   - The initiator calls `create_conversation` with the peer's intro bundle and a plain-text opening message.

>    - The recipient does not call anything; a `conversation_created` push event (`is_outgoing == false`) arrives automatically, followed by a `message_received` event.

Reviewer: The whole point here is two instances talking to each other, but the tutorial stops at "send a message" without showing the reader how to run two instances, swap intro bundles, and see a message land. `logos-chat-ui` already has this (`nix run .#exchange` / `two-instance-exchange.md`). Even a short "run a second instance, paste its bundle, you should see a `message_received` event" section — or just linking that flow as the verification step — would close the loop.

   ```

   - Message content is plain text in both directions — the module handles encoding and end-to-end encryption on the wire.

1. Read history and conversation state at any time (synchronous reads):

   ```cpp
   const QVariantList convos = m_logos->chat_module.list_conversations();  // [Conversation]
   const QVariantList msgs   = m_logos->chat_module.get_messages(convoId); // [Message]
   const QVariantMap  st     = m_logos->chat_module.status().toMap();      // { convo_count, delivery_state, detail }
   ```

   {% hint style="warning" %}
   Do not make a synchronous module read (`list_conversations`, `get_messages`, `status`) from *inside* an event handler — it re-enters the IPC replica while its read notifier is disabled and stalls until the call times out. Defer the read to the next event-loop turn instead (see `deferToEventLoop` in [`logos-chat-ui`](https://github.com/logos-co/logos-chat-ui/blob/master/src/ChatBackend.cpp)).
   {% endhint %}

1. Shut down cleanly:

   ```cpp
   m_logos->chat_module.shutdown();   // disconnects and tears the client down
   ```

## Step 5: Build and run

1. Build the module:

   ```sh
   nix build
   ```

> 1. Preview the module using `logos-standalone-app` (for `ui_qml` modules):

Reviewer: Step 5 says to preview the module in `logos-standalone-app`, but since we scaffold the backend template, delete the example UI, and only wire up the C++ chat logic, there's nothing visible to actually preview. Might be worth a line in "What to expect" saying this is the backend pattern and pointing at `logos-chat-ui` for a full wired UI, so nobody's surprised when `nix run` doesn't show a chat window.

---
title: Build a Logos module that uses the Chat module API
doc_type: procedure
product: messaging
topics: []
steps_layout: sectioned

> authors:

Reviewer 2: Still open from last round: `topics` and `authors` are empty, while the sibling Delivery doc fills them in. Filling both (adjust the author list as you see fit):

```suggestion
topics: chat
steps_layout: sectioned
authors: kashepavadan, igor-sirotin
```

- An understanding of [Logos modules](../../core/build-modules/build-a-logos-cpp-ui-module.md)

## What to expect

- You can initialise a chat client, connect to the Logos network, and exchange end-to-end encrypted messages with another instance.
- You can open a private 1:1 conversation by exchanging introduction bundles out of band and calling `create_conversation` from the initiating side.
- You can integrate the full chat lifecycle — init, subscribe to events, create bundle, open conversation, send, shut down — into any Logos C++ module.

## Step 1: Scaffold a new Logos module

Scaffold a new module using [`logos-module-builder`](https://github.com/logos-co/logos-module-builder). For a full walkthrough, see the [Build a Logos C++ UI module](../../core/build-modules/build-a-logos-cpp-ui-module.md) tutorial.

1. Create and enter the project directory:

   ```bash
   mkdir your-module-name && cd your-module-name
   ```

1. Initialise from the template:

   ```bash
   nix flake init -t github:logos-co/logos-module-builder/tutorial-v3#ui-qml-backend
   ```

1. Initialise a Git repository and stage all generated files:

   ```bash
   git init && git add -A
   ```

> 1. Remove the template's example sources. The scaffolded template includes `ui_example` files with mismatched class names and IIDs; leaving them causes build errors or plugin-load failures at runtime:

Reviewer 2: This step is where the happy path breaks. Deleting these files while `CMakeLists.txt` (`REP_FILE` / `SOURCES`) and `metadata.json` (`name`, `main`) still reference `ui_example` makes `nix build` fail at configure. The step needs to also cover re-pointing those two files and adding your own plugin sources:

```suggestion
1. Replace the template's example sources with your own. The scaffold's `CMakeLists.txt` (`REP_FILE` / `SOURCES`) and `metadata.json` (`name`, `main`) both reference `ui_example`, so simply deleting these files makes `nix build` (Step 5) fail at CMake configure with `file failed to open for reading: src/ui_example.rep`. Delete the examples, then create your own `src/<name>.rep`, `src/<name>_interface.h`, and `src/<name>_plugin.{h,cpp}` (Step 3 supplies the plugin body) and update `CMakeLists.txt` and `metadata.json` (`name`, `main`) to point at them:
```

Author: @igor-sirotin Is there an easier way to deal with this issue?

Reviewer 3: @weboko thanks for the catch!
@kashepavadan I've addressed this in c49e216fee9bddd8cf98f2c7609b98b8f7061421, please take a look

|---|---|---|
| `instance_path` | string | Directory for this instance's persistent state. Use a distinct directory per instance to run several side by side. |
| `delivery_preset` | string | Network preset for the delivery node. Use `logos.test` to reach the Logos test network. Must match across all participants. |
| `tcp_port` | int | Logos Delivery TCP port. `0` picks a random port. |

1. Initialise the chat client:

   ```cpp
   const LogosResult res = m_logos->chat_module.init("/path/to/instance", "logos.test", 0);
   if (!res.success) {
       qWarning() << "init failed:" << res.getError<QString>();
       return;
   }
   // init succeeded; delivery connects asynchronously.
   // Wait for delivery_state_changed with state == "online" before creating conversations.
   ```

2. Read (or set) your identity:

   ```cpp
   const QString myId = m_logos->chat_module.get_installation_name();
   // Optionally choose a name: m_logos->chat_module.set_installation_name("alice");
   ```

3. Create and share your introduction bundle:

   ```cpp
   const LogosResult bundle = m_logos->chat_module.create_intro_bundle();
   if (bundle.success)
      // share `bundleString` elsewhere

>       const auto myBundle = bundle.getValue<QString>()

Reviewer 2: This snippet won't compile as written — missing semicolon after `getValue<QString>()`, no braces so `myBundle` falls out of scope immediately, and the comment refers to `bundleString` but the variable is `myBundle`. Every other snippet in the doc uses the braced check-then-use form (matching `ChatBackend.cpp`), so let's match it here:

```suggestion
   const LogosResult bundle = m_logos->chat_module.create_intro_bundle();
   if (bundle.success) {
       const QString myBundle = bundle.getValue<QString>();
       // share `myBundle` out of band (the recipient pastes it — see Step 6)
   }
```

   ```

   - Message content is plain text in both directions — the module handles encoding and end-to-end encryption on the wire.

6. Read history and conversation state at any time (synchronous reads):

   ```cpp
   const QVariantList convos = m_logos->chat_module.list_conversations();  // [Conversation]
   const QVariantList msgs   = m_logos->chat_module.get_messages(convoId); // [Message]
   const QVariantMap  st     = m_logos->chat_module.status().toMap();      // { convo_count, delivery_state, detail }
   ```

   {% hint style="warning" %}
   Do not make a synchronous module read (`list_conversations`, `get_messages`, `status`) from *inside* an event handler — it re-enters the IPC replica while its read notifier is disabled and stalls until the call times out. Defer the read to the next event-loop turn instead (see `deferToEventLoop` in [`logos-chat-ui`](https://github.com/logos-co/logos-chat-ui/blob/v0.1.2/src/ChatBackend.cpp)).
   {% endhint %}

7. Shut down cleanly:

   ```cpp
   m_logos->chat_module.shutdown();   // disconnects and tears the client down
   ```

## Step 5: Build and run

1. Build the module:

   ```sh
   nix build
   ```

> 1. Preview the module using `logos-standalone-app` (for `ui_qml` modules):

Reviewer 2: Still open from last round: since Step 1 deletes the example UI and this guide only wires the C++ chat backend, `nix run` launches the standalone host with **no chat window** — which surprises readers who expect a visible chat. A one-line heads-up here (or in *What to expect*) closes that gap:

```suggestion
1. Preview the host with `logos-standalone-app` (for `ui_qml` modules). Because Step 1 removed the example UI and this guide only wires up the C++ chat backend, `nix run` starts the host but shows no chat window — that's expected. For a fully wired reference UI, see [`logos-chat-ui`](https://github.com/logos-co/logos-chat-ui/tree/v0.1.2):
```

- You can integrate the full chat lifecycle — init, subscribe to events, create bundle, open conversation, send, shut down — into any Logos C++ module.

## Step 1: Scaffold a new Logos module

Scaffold a new module using [`logos-module-builder`](https://github.com/logos-co/logos-module-builder). For a full walkthrough, see the [Build a Logos C++ UI module](../../core/build-modules/build-a-logos-cpp-ui-module.md) tutorial.

1. Create and enter the project directory:

   ```bash
   mkdir your-module-name && cd your-module-name
   ```

1. Initialise from the template:

   ```bash
   nix flake init -t github:logos-co/logos-module-builder/tutorial-v3#ui-qml-backend
   ```

1. Initialise a Git repository and stage all generated files:

   ```bash
   git init && git add -A
   ```

1. Remove the template's example sources. The scaffolded template includes `ui_example` files with mismatched class names and IIDs; leaving them causes build errors or plugin-load failures at runtime:

   ```bash
   rm -f src/ui_example.rep src/ui_example_interface.h src/ui_example_plugin.h src/ui_example_plugin.cpp

>    ```

Reviewer 2: **(blocker, reproduced) Following this step literally makes `nix build` fail at the CMake configure step.**

`tutorial-v3#ui-qml-backend` scaffolds a *coherent, buildable* plugin — `src/ui_example.rep` + `src/ui_example_plugin.*` — and both `CMakeLists.txt` (`REP_FILE`/`SOURCES`, line 21) and `metadata.json` (`main`) reference it. The `rm` here deletes those files but leaves the references, so `nix build` dies at configure:

```
CMake Error at .../cmake/LogosModule.cmake:602 (file):
  file failed to open for reading (No such file or directory):
    .../src/ui_example.rep
Call Stack (most recent call first):
  .../LogosModule.cmake:577 (_logos_module_add_replica_factory)
  CMakeLists.txt:21 (logos_module)
-- Configuring incomplete, errors occurred!
```

I verified the easiest fix is to **keep** the scaffolded plugin and wire the chat logic into its existing `initLogos()` in Step 3. With the files kept + the Step 2 dependencies added:
- `nix build` → produces `ui_example_plugin.so`
- `nix build .#lgx` (Step 5) → produces the `.lgx`
- adding the Step 3/4 `LogosModules` / `chat_module` calls into `initLogos()` compiles and links cleanly against the generated `logos_sdk.h` (`[13/16] ui_example_plugin.cpp` → `[16/16] Linking ui_example_plugin.so`, exit 0).

Also, the current rationale ("mismatched class names and IIDs; leaving them causes build errors or plugin-load failures") doesn't hold for `tutorial-v3` — the example plugin's class names and IID (`org.logos.UiExampleInterface`) are internally consistent and build fine. It's the deletion, not the example files, that breaks the build.

Suggested replacement for this step:

```suggestion
1. **Keep the scaffolded plugin — don't delete it.** In `tutorial-v3#ui-qml-backend` the generated `src/ui_example.rep` and `src/ui_example_plugin.*` are a coherent, buildable plugin that `CMakeLists.txt` (`REP_FILE`/`SOURCES`) and `metadata.json` (`main`) already reference; you'll wire the chat logic into its `initLogos()` in Step 3. Deleting these files without also re-pointing `CMakeLists.txt` and `metadata.json` breaks `nix build` at the CMake configure step (`file failed to open for reading … src/ui_example.rep`). To give the module its own name, **rename** the four `src/ui_example*` files and update the matching `REP_FILE`/`SOURCES` in `CMakeLists.txt` and `main` in `metadata.json` to match — don't remove them outright.
```

       flakeInputs = { delivery_module = logos-delivery-module; } // inputs;
     };
   ```

{% hint style="info" %}
The `dependency_overrides` entry above only points the builder at delivery's `.lidl` contract for code generation — it does **not** supply delivery's runtime build. The builder resolves each `metadata.json` dependency's runtime from a matching flake input, so `delivery_module` needs the `logos-delivery-module` input, mapped in via `flakeInputs`. Without it, `nix build` cannot resolve `delivery_module`. This mirrors how [`logos-chat-ui`](https://github.com/logos-co/logos-chat-ui/blob/v0.1.2/flake.nix) wires the two modules together.
{% endhint %}

## Step 3: Initialise `LogosModules` and subscribe to events

In your module's `initLogos()` function, construct `LogosModules` with the provided `LogosAPI*` and subscribe to all push events before calling `init()`. Subscribing first ensures you do not miss early events or the first incoming messages.

1. Initialise `LogosModules` in `initLogos()`:

   ```cpp
   #include "logos_sdk.h"   // generated umbrella — exposes LogosModules

   // Add a member to your plugin class (src/ui_example_plugin.h):

>    //   LogosModules* m_logos = nullptr;

Reviewer 2: **Build-breaking as written.** The `#include` belongs in the header, not the `.cpp` — the `LogosModules* m_logos` member is declared in `src/ui_example_plugin.h`, so the type must be visible there. As written the header fails with `'LogosModules' does not name a type`. Verified: moving the include into the header makes `nix build` pass.

```suggestion
   // Put this include in your plugin **header** (`src/ui_example_plugin.h`),
   // next to the `m_logos` member below — `LogosModules` must be a complete
   // type where the member is declared. The `.cpp` picks it up transitively
   // via its `#include "ui_example_plugin.h"`. Including it only in the `.cpp`
   // leaves the header failing with `error: 'LogosModules' does not name a type`.
   #include "logos_sdk.h"   // generated umbrella — exposes LogosModules

   // Add a member to your plugin class (src/ui_example_plugin.h):
   //   LogosModules* m_logos = nullptr;
```

   - `res.success` tells you whether the call succeeded;
   - `res.getError<QString>()` carries the failure reason;
   - `res.getValue<QString>()` carries the returned value (for example, the intro bundle).

Ongoing activity — incoming messages, new conversations, delivery-state changes — arrives **asynchronously** through the push events you subscribed to in Step 3.

{% hint style="info" %}
`init()` starts delivery asynchronously, so the client is not connected the moment `init()` returns. Watch `delivery_state_changed` for the `online` state before creating conversations or sending messages.
{% endhint %}

`init()` takes the instance directory, a delivery preset, and a TCP port:

| Parameter | Type | Notes |
|---|---|---|
| `instance_path` | string | Directory for this instance's persistent state. Use a distinct directory per instance to run several side by side. |
| `delivery_preset` | string | Network preset for the delivery node. Use `logos.test` to reach the Logos test network. Must match across all participants. |
| `tcp_port` | int | Logos Delivery TCP port. `0` picks a random port. |

1. Initialise the chat client:

   ```cpp

>    const LogosResult res = m_logos->chat_module.init("/path/to/instance", "logos.test", 0);

Reviewer 2: Dogfooding **Step 6** surfaced a gap that traces back to this line: it hardcodes the instance directory inside `initLogos()`, but Step 6 asks the reader to run *two* instances *"each with its own `instance_path`"*. Two `nix run` launches both write to the same `/path/to/instance`, sharing one `identity.db` / `history.json` — so they come up as the **same** identity instead of two peers, and the exchange can't happen. (`tcp_port: 0` is fine — each already gets a random port.)

Reading the directory from an env var makes Step 6 actually performable — launch the two instances with different `CHAT_INSTANCE_DIR` values:

```suggestion
   const QString dir = qEnvironmentVariable("CHAT_INSTANCE_DIR", "/tmp/chat-instance");
   const LogosResult res = m_logos->chat_module.init(dir, "logos.test", 0);
```

I verified this variant compiles against `chat_module` v0.1.2.

   const QVariantList msgs   = m_logos->chat_module.get_messages(convoId); // [Message]
   const QVariantMap  st     = m_logos->chat_module.status().toMap();      // { convo_count, delivery_state, detail }
   ```

   {% hint style="warning" %}
   Do not make a synchronous module read (`list_conversations`, `get_messages`, `status`) from *inside* an event handler — it re-enters the IPC replica while its read notifier is disabled and stalls until the call times out. Defer the read to the next event-loop turn instead (see `deferToEventLoop` in [`logos-chat-ui`](https://github.com/logos-co/logos-chat-ui/blob/v0.1.2/src/ChatBackend.cpp)).
   {% endhint %}

7. Shut down cleanly:

   ```cpp
   m_logos->chat_module.shutdown();   // disconnects and tears the client down
   ```

## Step 5: Build and run

1. Build the module:

   ```sh
   nix build
   ```

1. Preview the module using `logos-standalone-app` (for `ui_qml` modules):

   ```sh
   nix run                # preview via logos-standalone-app (for ui_qml modules)
   nix build .#lgx        # package as .lgx for installation into logos-basecamp
   ```

## Step 6: Verify a two-instance exchange

> A chat is only proven end to end when a message travels between two running instances. Start two copies of your module — each with its own `instance_path` and `tcp_port` — and confirm a message lands as a `message_received` event.

Reviewer 2: As written, Step 6 can't be run with the tutorial's own code, because `instance_path` is fixed in `initLogos()` (see the suggestion on the `init()` line) — there's no documented way to give two `nix run` instances distinct directories. Worth either parameterizing `instance_path` per the suggestion above, or pointing readers at the ready-made two-instance harness in [`logos-chat-ui`](https://github.com/logos-co/logos-chat-ui) (`nix run .#exchange`).
